core/dbt/task/run.py

```diff
@@ -284,18 +374,95 @@ def execute(self, model, manifest):
             )
 
         hook_ctx = self.adapter.pre_model_hook(context_config)
+        batch_results = None
         try:
-            result = MacroGenerator(
-                materialization_macro, context, stack=context["context_macro_stack"]
-            )()
+            if (
+                os.environ.get("DBT_EXPERIMENTAL_MICROBATCH")
+                and model.config.materialized == "incremental"
+                and model.config.incremental_strategy == "microbatch"
+            ):
+                batch_results = self._execute_microbatch_materialization(
+                    model, manifest, context, materialization_macro
+                )
+            else:
+                result = MacroGenerator(
+                    materialization_macro, context, stack=context["context_macro_stack"]
+                )()
+                for relation in self._materialization_relations(result, model):
+                    self.adapter.cache_added(relation.incorporate(dbt_created=True))
         finally:
             self.adapter.post_model_hook(context_config, hook_ctx)
 
-        for relation in self._materialization_relations(result, model):
-            self.adapter.cache_added(relation.incorporate(dbt_created=True))
+        if batch_results:
+            return self._build_run_microbatch_model_result(model, batch_results)
 
         return self._build_run_model_result(model, context)
 
+    def _execute_microbatch_materialization(self, model, manifest, context, materialization_macro):
```

Comment:
I think we can safely add typing to this function 🤔 But also maybe that's a lot harder than I think in this case 😅

Reply:
Done! 🙂